strawberry/extensions/apollo_cache_control.py

```diff
@@ -0,0 +1,179 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, Optional, Tuple, Union
+
+from graphql import GraphQLResolveInfo
+from graphql.pyutils.path import Path
+
+from strawberry.apollo.schema_directives import CacheControl, CacheControlScope
+from strawberry.extensions.base_extension import Extension
+from strawberry.field import StrawberryField
+from strawberry.types import ExecutionContext
+from strawberry.utils.await_maybe import AwaitableOrValue
+
+
+if TYPE_CHECKING:
+    from strawberry.schema.schema import Schema
+
+
+def is_cache_control_directive(directive) -> bool:
+    return isinstance(directive, CacheControl)
+
+
+@dataclass
+class CachePolicy:
+    max_age: Optional[int] = None
+    scope: Optional[CacheControlScope] = None
+
+    @classmethod
+    def from_directive(cls, directive: Optional[CacheControl] = None):
+        cache_policy = cls()
+        if directive:
+            if directive.max_age:
+                cache_policy.max_age = directive.max_age
+
+            if directive.scope:
+                cache_policy.scope = directive.scope
+
+        return cache_policy
+
+    def restrict(self, hint: CachePolicy):
+        if hint.max_age is not None and (
+            self.max_age is None or hint.max_age < self.max_age
+        ):
+            self.max_age = hint.max_age
+
+        if hint.scope is not None and self.scope != CacheControlScope.PRIVATE:
+            self.scope = hint.scope
+
+    @property
+    def policy(
+        self,
+    ) -> Dict[str, Union[int, str]]:
+        if self.max_age is None or self.max_age == 0:
+            return {}
+
+        return {
+            "max_age": self.max_age if self.max_age else 0,
+            "scope": self.scope.name.lower()
+            if self.scope
+            else CacheControlScope.PUBLIC.name.lower(),
+        }
+
+
+class ApolloCacheControlExtension(Extension):
+    default_max_age: int = 0
+
+    def __init__(
+        self,
+        *,
+        default_max_age: Optional[int] = 0,
+        calculate_http_headers: Optional[bool] = True,
+        execution_context: Optional[ExecutionContext] = None,
+    ):
+        self.calculate_http_headers = calculate_http_headers
+
+        if default_max_age:
+            self.default_max_age = default_max_age
+
+        if execution_context:
+            self.execution_context = execution_context
+
+    def on_request_start(self):
+        self.fields_caches: Dict[str, CachePolicy] = {}
+        self.overall_cache_policy = CachePolicy()
+
+    def on_executing_end(self):
+        if not self.calculate_http_headers:
+            return
+
+        self.execution_context.context["response"].headers[
```

Comment:
This should be tested for each integration... That means I should add ApolloCacheControlExtension to each schema... will mess up the tests I think, but I am not sure how to test this __in isolation__ otherwise...